### ISSUE TITLE
Halloy freezes and uses 100% of one CPU thread after it's left minimized for several hours

### ISSUE THREAD

**Issue by zsrv** [squidowl/halloy#589]
Halloy freezes and uses 100% of one CPU thread for up to a few minutes when I unminimize its window after leaving it minimized for several hours. During this time, Halloy is mostly unresponsive, but I will see old messages slowly appear in the channel I currently have open. After up to a few minutes, the CPU usage returns to normal and Halloy becomes responsive again. If I leave Halloy minimized for 30 minutes or so then unminimize the window, the problem will reoccur.

OS: openSUSE Tumbleweed
Halloy version: 2024.12 from the openSUSE Tumbleweed package repositories
If it matters, I am in ~50 channels. As of this moment, Halloy is using 530MB of memory.

**Comment by Daxot** [comment]
Same for me on sway. 
When halloy on another workspace that unfocused it's kinda "freezes in time" and after an hour or two when app gets focused again it speeds up to present time: it gets very laggy, all new messages in channel appears one after another, users in users channel list join and leave and so on.
Don't think that is related to channel count, because i'm in 12 channels right now.

OS: Void Linux
Halloy version: 2024.12

**Comment by casperstorm** [comment]
@tarkah @andymandias Any of you noticed this?

**Comment by andymandias** [comment]
I have seen this recently, as I've had Halloy open for longer periods of time.  I suspect this is a consequence of the solution for #364;  while in the background events queue up to be resolved, then the queue is processed once Halloy is foregrounded again.  I've only had to wait ~10s at the longest, but that's still slower than I would otherwise expect (you can watch the processing take place, with enough time to see events unfold).  My expectation is that this behavior is Wayland specific.

**Comment by tarkah** [comment]
Yeah seems winit has a bug for Wayland which causes the event loop to suspend. We keep processing messages in the background and queue them in a channel, and as soon as Halloy is brought back to the foreground, winit resumes and has to process all those messages.

**Comment by wezm** [comment]
>  My expectation is that this behavior is Wayland specific.

> Yeah seems winit has a bug for Wayland which causes the event loop to suspend.

I'm not sure this issue is limited to Wayland. I see the same thing on my X11 based Linux system (I'm only in 3 low traffic channels). I've started up a build with debug symbols and will poke at it in gdb if it does it again.

**Comment by Daxot** [comment]
After [7333371](https://github.com/squidowl/halloy/commit/733337195cecdfad783b007aa07a4b1adfd1f9a2) commit everything works great!
